### PR TITLE
S3900: Initial scaffolding

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Task 1, 2, 3 of #6793

### CODE REVIEW

**Comment by antonioaversa** [comment]
@zsolt-kolbay-sonarsource You may want to have a look at ITs failing: https://dev.azure.com/sonarsource/DotNetTeam%20Project/_build/results?buildId=63636&view=logs&j=c7ac5d00-ca19-592b-a6ee-71358b142e5c&t=d7fd8ed3-4895-531e-2ba8-7ed195398253&l=30386

It seems that expected results in ITs are not up-to-date.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+
+            b.ToString(); // Compliant, bool cannot be null
+
+            object v = null;
+            v.ToString(); // Compliant, we don't care about local variables
+
+            field.ToString(); // Compliant
+
+            Program.staticField.ToString(); // Compliant
+        }
```
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs:73]: Some other interesting scenarios, which don't seem to be currently covered:
- transitivity of nullable and non-nullable constraints on comparison => given that `o1 != null`, if you compare it against an object `o5` with no constraints, constraints propagate from `o1` to `o5`;
- transitivity of  nullable and non-nullable constraints on assignment => same as case above, but with an assignment: if you assign `o1` to `o5`, constraints propagage from `o1` to `o5`.

**Review by antonioaversa — CHANGES_REQUESTED** [review]
Partial review, reporting comments I had pending.
The rest of the review will happen once current issues with the PR are fixed and I am reassigned back as a reviewer.

File: analyzers/rspec/cs/S3900_c#.html [hunk 2]
```diff
@@ -1,5 +1,5 @@
-<p>A publicly accessible method can be called from anywhere, which means you should validate parameters to be within the expected constraints. In
-general, checking against <code>null</code> is recommended defensive programming.</p>
+<p>Methods declared as public, protected, or protected internal can be accessed from other assemblies, which means you should validate parameters to
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/rspec/cs/S3900_c#.html]: Added the formatting.

File: analyzers/rspec/cs/S3900_c#.html [hunk 3]
```diff
@@ -43,26 +33,42 @@ <h2>Compliant Solution</h2>
...
         {
             this.other = other.Clone();
         }
     }
+
+    public void Foo(MyOtherClass other)
+    {
+        this.other = other; // Compliant: "other" is not being dereferenced
+    }
 }
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/rspec/cs/S3900_c#.html:61]: Added a sample with a private method.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs [hunk 4]
```diff
@@ -0,0 +1,609 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs:1]: It is. 
e.g. `ArgumentException` and `Guid` are both used in the code and they reside in the `System` namespace.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+
+namespace Tests.Diagnostics
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs]: I reorganized the unit tests:
- there was a large namespace consisting of only C# 8.0 test code. I moved that to a separate file
- the namespaces were removed

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs [hunk 6]
```diff
@@ -0,0 +1,609 @@
...
+            o.ToString(); // FIXME Non-compliant {{Refactor this method to add validation of parameter 'o' before using it.}}
+
+            Bar(o); // Compliant, we care about dereference only
+
+            throw e; // FIXME Non-compliant
+        }
+
+        public void Bar(object o) { }
+
+        protected void NotCompliantCases_Nonpublic(object o)
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs]: Changed it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs [hunk 7]
```diff
@@ -0,0 +1,609 @@
...
+        private object field = null;
+        private static object staticField = null;
+
+        public void NotCompliantCases(object o, Exception e)
+        {
+            o.ToString(); // FIXME Non-compliant {{Refactor this method to add validation of parameter 'o' before using it.}}
+
+            Bar(o); // Compliant, we care about dereference only
+
+            throw e; // FIXME Non-compliant
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs]: I imagine certain fields or properties of the Exception (sub)class is being used when it's thrown. Essentially dereferenced.
I didn't see anything in the IL code, but the fact that it throws `NullReferenceException` implies that it's being dereferenced.
I added some explanation/comment.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs (hunk 1 above)
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs:86]: Those are really good test cases. Added them.

File: analyzers/rspec/cs/S3900_c#.html [hunk 8]
```diff
@@ -1,5 +1,6 @@
-<p>A publicly accessible method can be called from anywhere, which means you should validate parameters to be within the expected constraints. In
-general, checking against <code>null</code> is recommended defensive programming.</p>
+<p>Methods declared as <code>public</code>, <code>protected</code>, or <code>protected internal</code> can be accessed from other assemblies, which
```
Comment (pavel-mikula-sonarsource) [inline analyzers/rspec/cs/S3900_c#.html:1]: When having this, the PR description should also have "Fixes #2643"

Comment (pavel-mikula-sonarsource) [reply analyzers/rspec/cs/S3900_c#.html:1]: I would probably create a separate PR for it, not to block this progress on wording around RSPEC

File: analyzers/rspec/cs/S3900_c#.html [hunk 9]
```diff
@@ -31,38 +27,55 @@ <h2>Compliant Solution</h2>
...
+    private void Xyzzy(MyOtherClass other)
+    {
+        this.other = other; // Compliant: method is not publicly accessible
+    }
 }
 </pre>
 <h2>Exceptions</h2>
-<p>To create a custom null validation method declare an attribute with name <code>ValidatedNotNullAttribute</code> and mark the parameter that is
-validated for null in your method declaration with it:</p>
+<p>Arguments validated for <code>null</code> via helper methods should be annotated with the <a
```
Comment (pavel-mikula-sonarsource) [inline analyzers/rspec/cs/S3900_c#.html:64]: This paragraph should be probably formatted as 3 bullet points. Explaining those 3 alternative.

Currently, it reads as Microsoft, crosslinked to Reshaper followed by some custom code. It's hard to keep track and know that they are in OR mode.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.CSharp11.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+
+namespace Tests.Diagnostics
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.CSharp11.cs:3]: Please extract this into separate PR as described on the issue itself. There's too much to review

Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.CSharp11.cs:3]: You can delete these namespaces on the fly, they are just noise
```suggestion
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
The commit history is a mess and I can't see the actual changes. We need to split it into 3 PRs anyway so you should rebase, reset and re-create new clean commits. 

I couldn't do a full review

**Comment by zsolt-kolbay-sonarsource** [comment]
I'm closing this one, and adding all changes as separate - and better organized - pull requests.
